Step2

Mo-bile opened the pull request:
안녕하세요? 리뷰 요청드립니다. 😊

좀 더 제대로 해보고자 이번 미션은
`기능목록 작성 -> 도메인 목록과 역할 명세 -> 클래스 다이어그램 작성`
실시하면서 꼼꼼하게 챙기게 되었습니다.

개발 완료 후 최대한 리팩터링 할 부분을 찾고 리팩터링하지만 마땅히 하기 어려운 부분도 있었습니다.

고민되는 부분은 2곳이었습니다. 관련 질문을 요청드립니다
## 질문
1. `WinnerCount`를 Process 객체의 멤버로 넣어야 할 것인가 고민됩니다.
- 그런데 `Process` 의 필드인 `Lottos`와  `WinnerCount` 와는 생명주기가 달라서 추가는 어렵더라고요
즉 `Lottos`는 첫번째 입력(금액 입력)에 생기지만, `WinnerCount` 두번째(당첨번호) 입력에 생깁니다.

- 리팩터링 필요를 느끼는 이유는`WinnerCount` 에 있는 `calculateTotalReturn()` 메서드에 인자로 pay를 넣는게 바람직한가 고민이 들기 때문입니다.
왜냐하면 `Process` 필드로 pay가 있는데 이것을 활용하지 못하고, controller(main 함수)에서 입력받은 pay를 다시 활용해서 인자로 넣기 때문입니다.

이런 경우에 어떻게 하는게 좋을까요?

2. `WinnerCount` 에 4개 필드 이렇게 유지하는게 좋을지, 혹은 `Match` enum 을 필드로 할지, 혹은 4개 각각 모두 별개의 객체로 분리할지 고민됩니다.


감사합니다 👍

Commit:
feat : 로또 1회 생성완료

Commit:
fix : lotto 발행 생성기 숫자 수정

Commit:
feat : 로또 일급컬렉션 객체 생성하여 로또 리스트 생성기능 개발 완료

Commit:
feat : 기본 금액 입력 및 로또 세팅완료

Commit:
feat : 기본 입출력 개발 완료

Commit:
feat : 로또 일치 갯수 기능 개발 완료

Commit:
feat : 로또 결과 기능 개발 완료

Commit:
fix : 당첨로또 찾기에 멤버로 계산하지 않은 부분 수정

Commit:
feat : 당첨번호 입력시 수익률 계산 기능 개발 완료

Commit:
doc : 기능목록 체크 완료

Commit:
fix : 수익률 계산시 정밀도 손실을 방지하기 위해 BIGDECIMAL 로 변경

Commit:
refactor
1. 출력방식 수정
2. 로또 번호 입력 후 처리 시 공백 제거 추가

Commit:
refactor :
1. enum 을 이용하여 일치 갯수 카운트
2. Winner을 WinnerCount 객체로 이름을 변경하여 역할을 드러나게 함

Commit:
refactor : Lottos 의 findWinners 에 count를 증가시키는 역할 WinnerCount 로 이관

Commit:
refactor :
switch 문을 제거하기 위해 Match 에 상수별 추상 메서드 구현

Commit:
refactor
1. calculateTotalReturn 에서 이익 게산 메서드 분리
2. 당첨값 입력처리 메서드 분리

Commit:
refactor : pay 로 인자 및 변수명 통일

Commit:
doc : 리팩터링 내용 정리

javajigi commented on src/main/java/lotto/domain/business/Process.java:
```diff
@@ -0,0 +1,49 @@
+import lotto.domain.model.Lottos;
+import lotto.domain.model.WinnerCount;
+
+public class Process {
```
Process라는 이름이 너무 일반적인 느낌이 들어 AI에게 피드백을 요청해 봄
이 객체의 구체적인 역할과 책임을 파악할 수 있도록 이름을 지어보면 어떨까?
AI 피드백 결과 공유해 봄 

---
🎲 로또 게임 흐름 관리 객체 이름 추천
🥇 LottoGame (가장 권장)
설명: 이 객체가 로또 게임 자체의 실행 컨텍스트를 나타낸다는 것을 가장 직관적으로 표현합니다. **"로또 게임을 진행한다"**라는 비즈니스 로직의 총괄 책임자 역할을 명확히 합니다.

적합성: 내부에서 pay를 관리하고, Lottos를 생성하며, showWinners와 같이 게임의 핵심 과정을 진행하는 메서드를 가지고 있어 이름과 책임이 완벽하게 일치합니다.

🥈 LottoService
설명: 객체가 **도메인 내의 특정 비즈니스 기능(Service)**을 제공한다는 의미를 전달합니다. main 메서드를 가진 Application 객체가 이 서비스를 호출하여 게임을 구동하는 형태에 적합합니다.

장점: 레이어드 아키텍처에서 비즈니스 로직을 담당하는 계층임을 암시합니다.

🥉 LottoHost 또는 LottoManager
설명: **게임의 관리자(Host, Manager)**로서 게임의 시작(generateLottos), 진행(showWinners)을 주도한다는 의미를 부여합니다.

장점: 객체의 능동적인 역할을 강조할 수 있습니다.

💡 Process 이름이 아쉬운 이유
Process는 **'과정'**이라는 뜻의 명사이지만, 소프트웨어 개발에서는 **'특정 행위를 처리하는 일반적인 작업 단위'**를 의미하여 구체적인 책임이 불분명해집니다. 이 클래스가 로또 도메인 내에서 무엇을 처리하는지 알기 위해서는 반드시 코드를 열어봐야 한다는 단점이 있습니다.

결론적으로, LottoGame으로 이름을 변경하여 객체의 명확한 도메인 책임을 부여하는 것을 강력히 추천합니다.

javajigi commented on src/main/java/lotto/domain/business/LottoGame.java:
```diff
@@ -0,0 +1,49 @@
+
+public class Process {
+    
+    private final int pay;
```
원시 값을 포장하는 규칙 적용해 본다.

javajigi commented on src/main/java/lotto/domain/business/Process.java:
```diff
@@ -0,0 +1,49 @@
+        return this.lottos.getLottoList();
+    }
+    
+    public WinnerCount showWinners(String winnerLottoNumber) {
```
Lotto와 같이 6개의 값을 가지는 객체인데 Lotto에 String을 인자로 받는 생성자 추가하면 어떨까?

javajigi commented on src/main/java/lotto/domain/model/Lotto.java:
```diff
@@ -0,0 +1,66 @@
+    }
+    
+    private static List<Integer> setUpLottoNumber(List<Integer> numberList) {
+        List<Integer> lottoNumbers = numberList.subList(0, 6);
```
List.subList(fromIndex, toIndex)의 반환값을 새로운 ArrayList로 복사하지 않고 **재사용(reuse)**할 경우 발생하는 가장 큰 문제점은 원본 리스트의 구조적 변경이 subList에 영향을 미치고, 반대로 subList의 구조적 변경도 원본 리스트에 영향을 미친다는 점입니다.

이는 메모리 절약 측면에서는 재사용의 장점이 있지만, **예상치 못한 동작(ConcurrentModificationException 포함)**을 유발하여 버그를 발생시키기 쉽습니다.

❌ subList 재사용 시의 주요 문제점
1. 원본 리스트의 구조적 변경 영향 💣
subList 메서드가 반환하는 List는 **원본 리스트의 뷰(View)**입니다. 즉, 별도의 메모리를 할당하는 독립적인 리스트가 아니라, 동일한 데이터를 참조합니다.

만약 subList를 생성한 후, 원본 numberList에 다음과 같은 **구조적 변경(Structural Modification)**이 발생하면 문제가 생깁니다:

요소 추가/제거: numberList.add(newValue) 또는 numberList.remove(index)

리스트 크기 변경: numberList.clear()

이러한 변경이 발생한 후 subList의 메서드(예: subList.get(), subList.size())를 호출하면, Java는 ConcurrentModificationException을 발생시킵니다. 이는 subList가 유효하지 않은 상태가 되었음을 알리는 보호 메커니즘입니다.

2. subList 변경의 원본 리스트 전파 🔄
반대로, subList 자체에 구조적 변경을 가하면 그 변경 사항은 즉시 원본 리스트에 반영됩니다.

예시: subList.clear()를 호출하면, **원본 numberList**에서 해당 범위의 요소(0부터 5 인덱스까지)가 모두 제거됩니다.

문제: 의도치 않게 원본 데이터가 손상될 수 있으며, 원본 리스트를 사용하는 다른 코드 부분에 심각한 영향을 미칠 수 있습니다.

3. 메모리 누수 위험성 (Memory Leak Potential) 💧
subList는 원본 리스트의 내부 배열을 참조하고 있기 때문에, subList 객체만 남아있고 원본 numberList가 더 이상 사용되지 않더라도, **subList가 가비지 컬렉션(GC)**되기 전까지는 원본 리스트 전체의 메모리가 해제되지 않을 수 있습니다.

특히 큰 리스트의 작은 일부만 subList로 사용하고 나머지를 잊어버린 경우, 원본 리스트가 차지하는 큰 메모리 영역이 계속 유지되는 **메모리 누수(Memory Leak)**처럼 동작할 수 있습니다.

💡 안전한 재사용 방법
이러한 문제를 피하고 안전하게 subList의 값을 사용하려면, 항상 subList를 반환받은 직후 새로운 독립적인 컬렉션으로 복사해야 합니다.

```java
// 안전한 사용 (새로운 ArrayList 생성)
List<Integer> safeSubList = new ArrayList<>(numberList.subList(0, 6));

// 이후 safeSubList는 numberList의 변경과 무관하게 독립적으로 동작합니다.
```

javajigi commented on src/main/java/lotto/domain/model/Lottos.java:
```diff
@@ -0,0 +1,40 @@
+    public WinnerCount findWinners(Lotto winnerLotto) {
+        WinnerCount winnerCount = new WinnerCount();
+        for(Lotto lotto: this.lottoList) {
+            Match match = Match.fromLottoNumber(lotto.findMatchNumbers(winnerLotto.getNumbers()));
```
```suggestion
            Match match = lotto.match(winnerLotto);
```
로또의 등수 구하는 로직을 Lotto가 담당하도록 구현하면 어떨까?
객체를 수동적인 존재가 아니라 능동적인 존재로 바라보기 위해 노력해 본다.

javajigi commented on src/main/java/lotto/domain/model/WinningResult.java:
```diff
@@ -0,0 +1,93 @@
+    private int threeMatch;
+    private int fourMatch;
+    private int fiveMatch;
+    private int sixMatch;
```
Map으로 구현해 보면 어떨까?

javajigi commented on src/main/java/lotto/domain/model/WinnerCount.java:
```diff
@@ -0,0 +1,93 @@
+    private int fiveMatch;
+    private int sixMatch;
+    
+    public WinnerCount() {
```
WinnerCount 이름도 AI를 통해 더 적합한 이름으로 바꿔보면 어떨까?
객체를 너무 수동적인 존재로 바라보는 느낌이 든다.

javajigi commented on src/test/java/lotto/domain/model/LottosTest.java:
```diff
@@ -0,0 +1,25 @@
+    @Test
+    void 당첨번호를_입력받으면_당첨번호와_몇개_당첨인지_알린다() {
+        Lottos lottos = new Lottos(List.of(
+            new Lotto(List.of(1, 2, 3, 4, 5, 6)), new Lotto(List.of(40, 41, 42, 43, 44, 45))));
```
```suggestion
            new Lotto(1, 2, 3, 4, 5, 6), new Lotto(40, 41, 42, 43, 44, 45)));
```
가변 인자를 가지는 생성자를 Lotto에 추가해 위와 같이 구현해 본다.

javajigi reviewed (changes requested):
자동차 경주 게임 처음 구현했을 때의 복잡도에 비해 로또 미션의 첫 번째 구현의 객체 설계 좋네요. 👍 
단,, 아직도 객체를 수동적인 존재로 바라보는 경향이 느껴지네요.
좀 더 능동적으로 바라보면서 더 많은 책임을 부여해 보면 좋겠습니다.
이름을 지을 때 AI에게 피드백을 받아보면 좋겠어요.

Q1. WinnerCount를 Process 객체의 멤버로 넣어야 할 것인가 고민됩니다.
A1. 좋은 선택은 아니라 생각해요.
가능하면 필드는 현재와 같이 유지한 상태에서 WinnerCount를 실행 결과로 반환받는 방식으로 구현할 것을 제안해 봅니다.

Q2. WinnerCount 에 4개 필드 이렇게 유지하는게 좋을지, 혹은 Match enum 을 필드로 할지, 혹은 4개 각각 모두 별개의 객체로 분리할지 고민됩니다.
A2. 피드백에 남겼어요.
Match enum을 프로그램 실행할 때 인스터스 변수가 하나로 고정되기 때문에 상수 값만 필드로 가져야 합니다. 각 등수별 당첨 수와 같이 변경이 가능한 값을 필드로 가질 경우 백엔드의 멀티 쓰레드 환경에서 문제가 생길 수 있으니 주의하면 좋겠습니다. 이번 기회에 enum에 대해 추가 학습할 것을 추천해요.

Commit:
refactor : 클래스명 수정
1. `Process` 에서 `LottoGame` 로 수정하여, 기존 클래스명에서 드러나는 불분명한 책임을 방지하고, 본 로또 게임 자체의 실행 컨텍스트 담당을 직관적으로 표현
2. `WinnerCount`를 `WinningResult` 로 변경하여 기존 Count라는 단순 값에대한 수동적인 의미가 있고, DTO 처럼 보이지만 실질적으로 비즈니스 로직이 있다. 결과를 가지는 객체

Commit:
refactor :
1. 입력으로 로또를 만들 때 그 과정을 Lotto 내부로 옮겨서 행동하게 변경
2. 생성방식 pay, Input 으로 메서드 추가 및 변경
3. 테스트를 위한 생성자 추가

Commit:
fix : subList() 반환을 새로운 ArrayList로 넣어서, 메소드의 반환이 새로운 ArrayList가 아닌 참조 반환으로 발생할 문제 해결

Commit:
refactor
1. lotto에서 match된 상수를 반환하는 역할을 맡는다
2. enum 에 Match 관련 테스트 추가

Commit:
refactor : 매치 카운트를 이관하면서 타입을 객체 자기자신으로 바꿔서 get() 제거

Commit:
refactor : 원시값을 포장하여서 Pay로 검증 및 비즈니스 로직 이관

Commit:
refactor :
`WinningResult`의 필드 int에서 Map으로 전환

Commit:
refactor :
1. `WinningResult` 내 하드코딩된 부분 반복문으로 제거
2. increaseCount로 통합하여 반복된 메서드 제거

Commit:
refacot : 주요 숫자 상수화

Commit:
refactor : domain의 모든 class를 record로 전환하여 불변성 드러냄

Mo-bile commented:
안녕하세요 모재영입니다 👍 
리뷰 요청드립니다.

개인적으로 최대한 리팩터링 해보았는데요
하드코딩식 4개 상수 나열을 반복문으로 변형하면서 4개인데 오버엔지니어링 아닌가 하는 생각도 들었습니다
하지만 연습과정이니 충분하다고 생각들기도 했습니다

질문 드려도될까요?
## 질문 1 : 기존 활용한 메서드 접근지정자 변경하나요? (public -> private)
기존에 있는 메서드를 활용하는 새로운 public 메서드를 만드는경우 기존 메서드는 public 을 유지하나요 혹은 private로 변경하나요?
만약 변경한다면 기존 public기준으로 만든 테스트 코드도 모두 변경이 필요할 것 같아서요

예 : `Lotto` 클래스
```java
    public Match match(Lotto winNumbers) {
        return Match.fromLottoNumber(findMatchCount(winNumbers.numbers));
    }
    
    public int findMatchCount(List<Integer> winNumbers) {
        int matchCount = INITIAL_MATCH_COUNT;
        for(Integer number: this.numbers) {
            if(winNumbers.contains(number)) {
                matchCount++;
            }
        }
        return matchCount;
    }
```

기존 `findMatchCount()` 가 public이고 이게 맞추어서 테스트 코드가 존재했습니다.
그런데 리팩터링 하면서 다른 객체와 협력은 `match()`가 하게되었습니다.
그렇다면 기존 테스트 코드를 변경하고 `findMatchCount` 는 내부로 private로 변경해야할까요?

## 질문2 : 순환참조 느낌이 나는 이부분을 어떻게 할까요?
`WinningResult` 클래스의 `increaseMatch()` 메서드는 `Match`의 `increase()`를 호출합니다.
이 추상메서드 `increase()`는 각 상수마다 오버라이딩 되어 메서드 내부에서는 다시 `winningResult.increaseCount()`를 호출하는데요

WinningResult -> Match -> winningResult

이렇게 각각 객체들이 서로 연결되어 의존되느것 같은데 괜찮을까요?
Match 가 enum이라 괜찮은것 같기도 하고... 궁금합니다

## 질문3 : `WinningResult`에서 하드코딩된 부분을 반복문으로 바꾼 iteratorMatchMap(),  profit() 이 부분 뭔가 오버엔지니어링 같은데요 괜찮을까요?
연습과정이라 생각하고 했는데 만들고 나니 하드코딩된 4줄을 반복문으로 바꾸니 3줄이 되는... 현상을 보고 아차 싶었기 때문입니다.

예
```java
// before
    private BigDecimal getProfit() {
        BigDecimal profit = BigDecimal.ZERO;
        profit = profit.add(BigDecimal.valueOf(this.matchMap.get(THREE_MATCH)).multiply(BigDecimal.valueOf(THREE_MATCH.getWinnerReturn())));
        profit = profit.add(BigDecimal.valueOf(this.matchMap.get(FOUR_MATCH)).multiply(BigDecimal.valueOf(FOUR_MATCH.getWinnerReturn())));
        profit = profit.add(BigDecimal.valueOf(this.matchMap.get(FIVE_MATCH)).multiply(BigDecimal.valueOf(FIVE_MATCH.getWinnerReturn())));
        profit = profit.add(BigDecimal.valueOf(this.matchMap.get(SIX_MATCH)).multiply(BigDecimal.valueOf(SIX_MATCH.getWinnerReturn())));
        return profit;
    }

// after
    private BigDecimal getProfit() {
        BigDecimal profit = BigDecimal.ZERO;
        for(Entry<Match, Integer> matchIntegerEntry: matchMap.entrySet()) {
            profit = profit.add(BigDecimal.valueOf(matchIntegerEntry.getValue()).multiply(BigDecimal.valueOf(matchIntegerEntry.getKey().getWinnerReturn())));
        }
        return profit;
    }
```


```java
// before
    private static Map<Match, Integer> initMatchMap(int threeMatchCount, int fourMatchCount, int fiveMatchCount, int sixMatchCount) {
        Map<Match, Integer> matchMap = new HashMap<>();
        matchMap.put(Match.THREE_MATCH, threeMatchCount);
        matchMap.put(Match.FOUR_MATCH, fourMatchCount);
        matchMap.put(Match.FIVE_MATCH, fiveMatchCount);
        matchMap.put(Match.SIX_MATCH, sixMatchCount);
        return matchMap;
    }
// after
    private static Map<Match, Integer> initMatchMap(int threeMatchCount, int fourMatchCount, int fiveMatchCount, int sixMatchCount) {
        Map<Match, Integer> matchMap = new HashMap<>();
        iteratorMatchMap(matchMap, List.of(threeMatchCount, fourMatchCount, fiveMatchCount, sixMatchCount));
        return matchMap;
    }
    
    private static void iteratorMatchMap(Map<Match, Integer> matchMap, List<Integer> count) {
        for(int i = 0; i < Match.values().length; i++) {
            matchMap.put(Match.values()[i], count.get(i));
        }
    }
```

감사합니다 👍

Commit:
refactor : 불필요한 `match` 내 increase 를 제거하고 자기 스스로를 직접 호출하게 하여 순환참조 방지

Mo-bile commented:
순환참조 질문 드리다가 문뜩 서로서로 this 인자로 전달하는 것을 보았습니다.
그래서 서로 호출을 끊어내고 `WinningResult` 에서 바로 호출하면 될것 같아서 바로 스스로 호출을 시키는 방식으로 리팩터링했습니다
참고부탁드리겠습니다 👍

javajigi commented on src/main/java/lotto/domain/business/LottoGame.java:
```diff
@@ -0,0 +1,26 @@
+        return this.lottos.lottoList();
+    }
+    
+    public WinningResult showWinners(String winnerLottoNumber) {
+        return this.lottos.findWinners(new Lotto(winnerLottoNumber));
+    }
```
AI에게 showWinners, findWinners이 적합한지 피드백 받아볼 것을 추천

javajigi commented on src/main/java/lotto/domain/model/Lotto.java:
```diff
@@ -0,0 +1,74 @@
+        this(generateLottoNumberByInput(lottoNumbers));
+    }
+    
+    public Lotto(int numA, int numB, int numC, int numD, int numE, int numF) {
```
```suggestion
    public Lotto(int... nums) {
```
가변인자로 구현 가능함
이번 기회에 가변인자 추가학습할 것을 추천

javajigi commented on src/main/java/lotto/domain/model/Pay.java:
```diff
@@ -0,0 +1,20 @@
+package lotto.domain.model;
+
+public record Pay(int pay) {
```
👍

javajigi commented on src/main/java/lotto/domain/model/WinningResult.java:
```diff
@@ -0,0 +1,71 @@
+        this(initMatchMap(INITIAL_MATCH_COUNT, INITIAL_MATCH_COUNT, INITIAL_MATCH_COUNT, INITIAL_MATCH_COUNT));
+    }
+    
+    public WinningResult(int threeMatchCount, int fourMatchCount, int fiveMatchCount, int sixMatchCount) {
```
이 생성자가 굳이 필요할까?

Mo-bile replied:
음 테스트 코드 때문에 필요하게 되었습니다 👍

javajigi commented on src/main/java/lotto/domain/model/WinningResult.java:
```diff
@@ -0,0 +1,71 @@
+        return profit;
+    }
+    
+    public void increaseMatch(Match match) {
```
increaseMatch, increaseCount와 같은 메서드 이름 관련해 ai에게 피드백 받아볼 것을 추천

javajigi commented on src/test/java/lotto/domain/model/LottoTest.java:
```diff
@@ -0,0 +1,64 @@
+    @Test
+    public void 인자로들어온_로또번호가_3개_일치한다() {
+        Lotto lotto = new Lotto(40, 41, 42, 43, 44, 45);
+        assertThat(lotto.findMatchCount(List.of(35, 36, 37, 43, 44, 45)))
```
```suggestion
        assertThat(lotto.findMatchCount(new Lotto(35, 36, 37, 43, 44, 45)))
```
Lotto를 사용해도 되지 않을까?

javajigi commented on src/main/java/lotto/domain/model/Lotto.java:
```diff
@@ -0,0 +1,74 @@
+    }
+    
+    public Match match(Lotto winNumbers) {
+        return Match.fromLottoNumber(findMatchCount(winNumbers.numbers));
```
```suggestion
        return Match.fromLottoNumber(winNumbers.findMatchCount(this.numbers));
```
위와 같이 구현할 수도 있지 않을까?

javajigi reviewed (approved):
피드백 반영하느라 수고했어요. 👍 
피드백 몇개 남겼는데요.
객체 설계에 큰 영향을 미치지 않는 피드백이라 바로 merge하니 3단계 진행할 때 함께 반영해 보세요.

Q1. 기존 활용한 메서드 접근지정자 변경하나요? (public -> private)
기존 findMatchCount() 가 public이고 이게 맞추어서 테스트 코드가 존재했습니다.
그런데 리팩터링 하면서 다른 객체와 협력은 match()가 하게되었습니다.
그렇다면 기존 테스트 코드를 변경하고 findMatchCount 는 내부로 private로 변경해야할까요?
A1. findMatchCount()의 값에 따른 Match를 구하는 로직을 테스트할 때 함께 테스트 가능하기 때문에 private으로 변경하고 굳이 테스트하지 않아도 되지 않아도 됨

Q2. 순환참조 느낌이 나는 이부분을 어떻게 할까요?
WinningResult 클래스의 increaseMatch() 메서드는 Match의 increase()를 호출합니다.
이 추상메서드 increase()는 각 상수마다 오버라이딩 되어 메서드 내부에서는 다시 winningResult.increaseCount()를 호출하는데요
A2. Match의 increase()를 호출하는 부분은 보이지 않는데요. 한번 확인해 보면 좋겠습니다.

Q3. WinningResult에서 하드코딩된 부분을 반복문으로 바꾼 iteratorMatchMap(), profit() 이 부분 뭔가 오버엔지니어링 같은데요 괜찮을까요?
A3. 피드백에 남겼는데요.
저는 WinningResult(int threeMatchCount, int fourMatchCount, int fiveMatchCount, int sixMatchCount) 생성자가 왜 필요한지 모르겠습니다. 제거해도 되지 않을까요?
WinningResult를 생성할 때 값이 Match에 해당하는 값이 0인 Map을 초기화하면 되지 않을까요?
